pkg/api/v1/conversion.go

```diff
@@ -555,109 +427,6 @@ func Convert_v1_PodSpec_To_api_PodSpec(in *v1.PodSpec, out *api.PodSpec, s conve
 	return nil
 }
 
-func Convert_api_Pod_To_v1_Pod(in *api.Pod, out *v1.Pod, s conversion.Scope) error {
-	if err := autoConvert_api_Pod_To_v1_Pod(in, out, s); err != nil {
-		return err
-	}
-
-	// TODO: sometime after we move init container to stable, remove these conversions
-	if len(out.Spec.InitContainers) > 0 || len(out.Status.InitContainerStatuses) > 0 {
-		old := out.Annotations
-		out.Annotations = make(map[string]string, len(old))
-		for k, v := range old {
-			out.Annotations[k] = v
-		}
-		delete(out.Annotations, v1.PodInitContainersAnnotationKey)
-		delete(out.Annotations, v1.PodInitContainersBetaAnnotationKey)
-		delete(out.Annotations, v1.PodInitContainerStatusesAnnotationKey)
-		delete(out.Annotations, v1.PodInitContainerStatusesBetaAnnotationKey)
-	}
-	if len(out.Spec.InitContainers) > 0 {
-		value, err := json.Marshal(out.Spec.InitContainers)
-		if err != nil {
-			return err
-		}
-		out.Annotations[v1.PodInitContainersAnnotationKey] = string(value)
-		out.Annotations[v1.PodInitContainersBetaAnnotationKey] = string(value)
-	}
-	if len(out.Status.InitContainerStatuses) > 0 {
-		value, err := json.Marshal(out.Status.InitContainerStatuses)
-		if err != nil {
-			return err
-		}
-		out.Annotations[v1.PodInitContainerStatusesAnnotationKey] = string(value)
-		out.Annotations[v1.PodInitContainerStatusesBetaAnnotationKey] = string(value)
-	}
-
-	return nil
-}
-
-func Convert_v1_Pod_To_api_Pod(in *v1.Pod, out *api.Pod, s conversion.Scope) error {
-	// If there is a beta annotation, copy to alpha key.
-	// See commit log for PR #31026 for why we do this.
-	if valueBeta, okBeta := in.Annotations[v1.PodInitContainersBetaAnnotationKey]; okBeta {
-		in.Annotations[v1.PodInitContainersAnnotationKey] = valueBeta
-	}
-	// TODO: sometime after we move init container to stable, remove these conversions
-	// Move the annotation to the internal repr. field
-	if value, ok := in.Annotations[v1.PodInitContainersAnnotationKey]; ok {
-		var values []v1.Container
-		if err := json.Unmarshal([]byte(value), &values); err != nil {
-			return err
-		}
-		// Conversion from external to internal version exists more to
-		// satisfy the needs of the decoder than it does to be a general
-		// purpose tool. And Decode always creates an intermediate object
-		// to decode to. Thus the caller of UnsafeConvertToVersion is
-		// taking responsibility to ensure mutation of in is not exposed
-		// back to the caller.
-		in.Spec.InitContainers = values
-		// Call defaulters explicitly until annotations are removed
-		tmpPod := &v1.Pod{
-			Spec: v1.PodSpec{
-				HostNetwork:    in.Spec.HostNetwork,
-				InitContainers: values,
-			},
-		}
-		SetObjectDefaults_Pod(tmpPod)
-		in.Spec.InitContainers = tmpPod.Spec.InitContainers
```

Comment:
@smarterclayton I hadn't realized that annotations were trumping the fields. this means that version skewed kubelets will honor annotations over fields, which means the server's view of what the pod will be doing (in terms of permissions, resources, images run, etc) and the kubelets can be different, and a user can target version skewed kubelets with init container annotations. should we actively clear the alpha/beta init container annotations in the apiserver? marking do-not-merge until we resolve this

Reply:
Can the masters remove them on the way out?

Reply:
@smarterclayton do you mean that we can just leave `Convert_v1_Pod_To_api_Pod` and `Convert_api_Pod_To_v1_Pod` as-is for one/two more release(s)?

Reply:
We probably want to always clear the annotations for beta and alpha when converting internal to v1.  That might confuse some agents, but is the correct behavior since fields have been available since 1.6

Reply:
I think that spot would work, but we'd want to drop the annotations unconditionally, maybe like this:

```
func Convert_v1_Pod_To_api_Pod(in *v1.Pod, out *api.Pod, s conversion.Scope) error {
	if err := autoConvert_v1_Pod_To_api_Pod(in, out, s); err != nil {
		return err
	}
	// drop init container annotations so they don't take effect on legacy kubelets.
	// remove this once the oldest supported kubelet no longer honors the annotations over the field.
	delete(out.Annotations, v1.PodInitContainersAnnotationKey)
	delete(out.Annotations, v1.PodInitContainersBetaAnnotationKey)
	delete(out.Annotations, v1.PodInitContainerStatusesAnnotationKey)
	delete(out.Annotations, v1.PodInitContainerStatusesBetaAnnotationKey)
	return nil
}
 
func Convert_api_Pod_To_v1_Pod(in *api.Pod, out *v1.Pod, s conversion.Scope) error {
	if err := autoConvert_api_Pod_To_v1_Pod(in, out, s); err != nil {
		return err
	}
	// drop init container annotations so they don't take effect on legacy kubelets.
	// remove this once the oldest supported kubelet no longer honors the annotations over the field.
	delete(out.Annotations, v1.PodInitContainersAnnotationKey)
	delete(out.Annotations, v1.PodInitContainersBetaAnnotationKey)
	delete(out.Annotations, v1.PodInitContainerStatusesAnnotationKey)
	delete(out.Annotations, v1.PodInitContainerStatusesBetaAnnotationKey)
	return nil
}
```